internal/transports/p2p/peer/peer.go

```diff
@@ -0,0 +1,558 @@
+		}
+	}
+
+	p.log.Info().Msgf("ping handler shutdown for peer %s", p)
+	p.wg.Done()
+}
+
+// MsgHandler is a message handler for incoming messages.
+// Must be run as a goroutine.
+func (p *Peer) readMsgHandler() {
+	p.wg.Add(1)
+
+out:
+	for {
+		select {
+		case <-p.quit:
+			break out
+
+		default:
+			remoteMsg, _, err := wire.ReadMessage(p.conn, p.protocolVersion, p.chainParams.Net)
+			if err != nil {
+				if !p.quitting {
+					p.log.Error().Msgf("cannot read message from peer %s, reason: %v", p, err)
+				}
+				continue
+			}
+
+			switch msg := remoteMsg.(type) {
+			case *wire.MsgPing:
+				p.handlePingMsg(msg)
+			case *wire.MsgPong:
+				p.log.Info().Msgf("received pong from peer %s with nonce: %d", p, msg.Nonce)
+			case *wire.MsgHeaders:
+				p.handleHeadersMsg(msg)
+			case *wire.MsgInv:
+				p.handleInvMsg(msg)
+			default:
+				p.log.Info().Msgf("received msg of type: %T", msg)
+			}
+		}
+	}
+
+	p.log.Info().Msgf("read msg handler shutdown for peer %s", p)
+	p.wg.Done()
+}
+
+func (p *Peer) writeMessage(msg wire.Message) error {
+	err := wire.WriteMessage(p.conn, msg, p.protocolVersion, p.chainParams.Net)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (p *Peer) writeRejectMessage(msg wire.Message, reason string) {
+	rejectMsg := wire.NewMsgReject(msg.Command(), wire.RejectObsolete, reason)
+	err := p.writeMessage(rejectMsg)
+	if err != nil {
+		p.log.Error().Msgf("could not write reject message to peer, reason: %v", err)
+	}
+}
+
+func (p *Peer) queueMessage(msg wire.Message) {
+	// running in goroutine here, because writing
+	// to channel is blocking until read
+	go func() {
+		p.msgChan <- msg
+	}()
+}
+
+// writeMsgHandler serves as a queue for writing messages to peers,
+// must be run as a goroutine.
+func (p *Peer) writeMsgHandler() {
+	p.wg.Add(1)
+
+out:
+	for {
+		select {
+		case msg := <-p.msgChan:
+			err := p.writeMessage(msg)
+			if err != nil {
+				p.log.Error().Msgf("error writing msg %T to peer %s", msg, p)
+			}
+		case <-p.quit:
+			break out
+		}
+	}
+
+cleanup:
+	for {
+		select {
+		case <-p.msgChan:
+		default:
+			break cleanup
+		}
+	}
+
+	p.log.Info().Msgf("write msg handler shutdown for peer %s", p)
+	p.wg.Done()
+}
+
+func (p *Peer) writeOurVersionMsg() error {
+	n, err := rand.Int(rand.Reader, big.NewInt(9223372036854775807))
+	if err != nil {
+		panic(err)
+	}
+	nonce := n.Uint64()
+	p.nonce = nonce
+
+	ourNA := &wire.NetAddress{
+		Services:  wire.SFspv,
+		Timestamp: time.Now(),
+	}
+	theirNA := wire.NewNetAddress(p.addr, 0)
+
+	lastBlock := p.headersService.GetTip().Height
+
+	msg := wire.NewMsgVersion(ourNA, theirNA, nonce, lastBlock)
+	err = msg.AddUserAgent(p.cfg.UserAgentName, p.cfg.UserAgentVersion, userAgentComments)
+	if err != nil {
+		p.log.Error().Msgf("could not add user agent to version message, reason: %v", err)
+		return err
+	}
+
+	// NOTE: it's 0 by default, so in theory we don't need
+	// to set that, but it's better to be explicit.
+	msg.Services = p.services
```

Comment:
I can't get the connection between the comment (about "0 by default") and the line below it "msg.Services = p.services"

Reply:
Because p.services is 0, and msg.Services is also 0 by default, so we don't need to set this line `msg.Services = p.services`, but I actually want to do that to be explicit in that case.

But you're right that this comment is misleading, removed :)